File: go/day04/main.go
package main

import (
	"fmt"
	"strings"

	utils "github.com/MaraudingAvenger/Advent2025/common"
)

func main() {
	lines, err := utils.ReadInputFile(utils.FindInputFile("inputs/day4.txt"))
	if err != nil {
		fmt.Println("Error reading input file:", err)
		return
	}
	var linesList [][]rune
	for _, line := range strings.Fields(lines) {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		linesList = append(linesList, []rune(line))
	}

	accessible := 0

	for i := 0; i < len(linesList); i++ {
		for j := 0; j < len(linesList[i]); j++ {
			if linesList[i][j] == '@' {
				// Count the number of '@' symbols in the surrounding cells
				atCount := 0
				for di := -1; di <= 1; di++ {
					for dj := -1; dj <= 1; dj++ {
						if di == 0 && dj == 0 {
							continue
						}
						ni, nj := i+di, j+dj
						if ni >= 0 && ni < len(linesList) && nj >= 0 && nj < len(linesList[ni]) {
							if linesList[ni][nj] == '@' {
								atCount++
							}
						}
					}
				}

				if atCount < 4 {
					accessible++
				}
			}
		}
	}
	fmt.Printf("Part 1: %d\n", accessible)

	// Part 2: Remove the '@' symbols that are accessible and loop again until no more can be removed
	accessible = 0

	removed := false
	for {
		for i := 0; i < len(linesList); i++ {
			for j := 0; j < len(linesList[i]); j++ {
				if linesList[i][j] == '@' {
					// Count the number of '@' symbols in the surrounding cells
					atCount := 0
					for di := -1; di <= 1; di++ {
						for dj := -1; dj <= 1; dj++ {
							if di == 0 && dj == 0 {
								continue
							}
							ni, nj := i+di, j+dj
							if ni >= 0 && ni < len(linesList) && nj >= 0 && nj < len(linesList[ni]) {
								if linesList[ni][nj] == '@' {
									atCount++
								}
							}
						}
					}

					if atCount < 4 {
						linesList[i][j] = '.'
						accessible++
						removed = true
					}
				}
			}
		}
		if !removed {
			break
		}
		removed = false
	}
	fmt.Printf("Part 2: %d\n", accessible)
}

File: rust/day04/src/main.rs
use common::read_lines;
use common::find_input_file;

fn count_at_symbols(grid: &[Vec<u8>], i: usize, j: usize) -> u8 {
    let rows = grid.len() as isize;
    let cols = grid[0].len() as isize;
    let mut at_count = 0;

    for di in -1..=1 {
        for dj in -1..=1 {
            if di == 0 && dj == 0 {
                continue;
            }
            let ni = i as isize + di;
            let nj = j as isize + dj;
            if ni >= 0 && ni < rows && nj >= 0 && nj < cols && grid[ni as usize][nj as usize] == b'@' {
                at_count += 1;
            }
        }
    }

    at_count
}

fn main() {
    let mut grid: Vec<Vec<u8>> = read_lines(&find_input_file("inputs/day4.txt"))
        .into_iter()
        .map(|line| line.into_bytes())
        .collect();

    let mut accessible = 0;
    for i in 0..grid.len() {
        for j in 0..grid[0].len() {
            if grid[i][j] == b'@' && count_at_symbols(&grid, i, j) < 4 {
                accessible += 1;
            }
        }
    }

    println!("Part 1: {}", accessible);

    accessible = 0;
    loop {
        let mut removed = false;
        for i in 0..grid.len() {
            for j in 0..grid[0].len() {
                if grid[i][j] == b'@' && count_at_symbols(&grid, i, j) < 4 {
                    grid[i][j] = b'.';
                    accessible += 1;
                    removed = true;
                }
            }
        }

        if !removed {
            break;
        }
    }

    println!("Part 2: {}", accessible);
}
